java/arcs/android/util/ProtoPrefetcher.kt

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+package arcs.android.util
+
+import android.os.Trace
+import arcs.android.crdt.CrdtCountProto
+import arcs.android.crdt.CrdtDataProto
+import arcs.android.crdt.CrdtEntityProto
+import arcs.android.crdt.CrdtEntityReferenceProto
+import arcs.android.crdt.CrdtOperationProto
+import arcs.android.crdt.CrdtSetProto
+import arcs.android.crdt.CrdtSingletonProto
+import arcs.android.crdt.RawEntityProto
+import arcs.android.crdt.ReferencablePrimitiveProto
+import arcs.android.crdt.ReferencableProto
+import arcs.android.crdt.ReferencableSetProto
+import arcs.android.crdt.ReferenceProto
+import arcs.android.crdt.VersionMapProto
+import arcs.android.storage.ProxyMessageProto
+import java.util.concurrent.Executors
+
+/**
+ * Run asynchronously as a cache warmer to suffer from Arcs protos initialization
+ * overhead as well as cache misses on behalf of whoever needs Arcs capabilities.
+ *
+ * The initialization overhead typically comes from instantiating default instance
+ * and parsing/building proto schema.
+ *
+ * Recommend executing [ProtoPrefetcher.prefetch] as early as possible i.e.
+ * the entry of Application lifecycle: `onCreate()`
```

Comment:
Might be worth looking into using the new androidx.startup tools to do this automatically:

https://android-review.googlesource.com/c/platform/frameworks/support/+/1206056

Reply:
Interesting, took a quick look at their usage and what was the problem they tried to solve.
They were trying to get rid of expensive ContentProvider code at app initialization by calling a single AppInitializer.initialize(...) at Application.onCreate() to trigger discovery and initialization of defined components.

How they use in Cloak:
"""
When enabled, Cloak could then inline the call to AppInitializer.initialize(...) in the generated Application.onCreate(), and the developer is also prompted to remove the default provider (which would have initialized components). This way, there is no need for any ContentProvider during Application.onCreate().
"""

ha, it's too fancy to me as I also wanted a single inline ProtoPrefetcher.prefetch() at Application.onCreate() and Service.onCreate().